### PR TITLE
Merge windowsai (winml layering) into master

### PR DESCRIPTION
**Pull request by smk2007**
This PR is for comment on bringing the newly layered Windows ML components into the master branch for the ONNX runtime for windows.

This is anticipation of a beta release of these bits over the next couple of months. We will include more documentation on how to use , how the layers work, and the relationships of WinML and ORT and DML (which we introduced to master in ORT v1.0) .

Some of the things we have done in this PR:
    - Added a top level directory "/winml"
    - Contributed all of the windows inbox code from the Windows.AI.MachineLearning namespace into that directory . Making it available using the MIT license.
    - Started a layering effort to have the new Windows.AI.MachineLearning.dll consume the onnxruntime.dll c_abi that we introduced in v1.0 .
    - Added an "adapter" module that gets linked into the core onnxruntime.dll. This adapter is private to the WinML component, and provides ABI functionality required for the layering effort. this is not a new public ABI and is not support for developer to call.
    - Made the WinML ABI fully available for public developers to call.
    - You can now include both of these DLL's (WinML + ORT) in your projects that want to use the WinML ABI as a redist component.
    - Added cmakery for all of this work. There is now a "use_winml" build flag in addition to the "use_dml" build flag
    - Added google test unit tests for the newly added WinML ABI. these are under the top level "winml/test" folder.
    - And several other things :)

Enjoy !

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -82,7 +82,9 @@ option(tensorflow_C_PACKAGE_PATH "Path to tensorflow C package installation dir"
 option(onnxruntime_ENABLE_LANGUAGE_INTEROP_OPS "Enable operator implemented in language other than cpp" OFF)
 option(onnxruntime_DEBUG_NODE_INPUTS_OUTPUTS "Dump node input shapes and output data to standard output when executing the model." OFF)
 option(onnxruntime_USE_DML "Build with DirectML support" OFF)
+option(onnxruntime_USE_WINML "Build with WinML support" OFF)
 option(onnxruntime_USE_ACL "Build with ACL support" OFF)
+option(onnxruntime_USE_TELEMETRY "Build with Telemetry" OFF)
```
Comment (snnn) [inline cmake/CMakeLists.txt]: Dup?

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -199,10 +201,12 @@ if (MSVC)
     SET (CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /Gw /GL")
     SET (CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} /Gw /GL")
   endif()
-  check_cxx_compiler_flag(-Qspectre HAS_QSPECTRE)
-  if (HAS_QSPECTRE)
-    SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /Qspectre")
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Qspectre")
+  if (NOT DEFINED onnxruntime_DISABLE_QSPECTRE_CHECK)
```
Comment (snnn) [inline cmake/CMakeLists.txt]: Please add a comment of why this special handling is needed.

File: onnxruntime/core/graph/model.cc [hunk 3]
```diff
@@ -95,6 +95,10 @@ Model::Model(std::unique_ptr<ModelProto> model_proto, const IOnnxRuntimeOpSchema
         " specifies which version of the ONNX OperatorSet is being imported.");
   }
 
+  if (!model_proto->has_ir_version() || model_proto->ir_version() > ONNX_NAMESPACE::Version::IR_VERSION) {
```
Comment (snnn) [inline onnxruntime/core/graph/model.cc]: Conservatively, I think we should follow the same approach for the opset version. That is, we should explicitly define the max ir version in our source code, instead of using a value from ONNX (ONNX_NAMESPACE::Version::IR_VERSION)).

File: onnxruntime/core/graph/schema_registry.cc [hunk 4]
```diff
@@ -194,6 +194,23 @@ DomainToVersionMap SchemaRegistryManager::GetLatestOpsetVersions(bool is_onnx_on
   return domain_version_map;
 }
 
+static bool IsDomainVersionBeyondSupportedRange(
```
Comment (snnn) [inline onnxruntime/core/graph/schema_registry.cc]: Does it have any impact on custom ops?

File: onnxruntime/core/platform/windows/debug_alloc.cc [hunk 5]
```diff
@@ -232,12 +232,8 @@ Memory_LeakCheck::~Memory_LeakCheck() {
...
     string.append(buffer);
 
-    // If we're being actively debugged, show a message box to get the dev's attention
-    if (IsDebuggerPresent())
-      MessageBoxA(nullptr, string.c_str(), "Warning", MB_OK | MB_ICONWARNING);
-    else {
-      // If we're on the command line (like on a build machine), output to the console and exit(-1)
-      std::cout << "\n----- MEMORY LEAKS: " << string.c_str() << "\n";
+    std::cout << "\n----- MEMORY LEAKS: " << string.c_str() << "\n";
+    if (!IsDebuggerPresent()) {
```
Comment (snnn) [inline onnxruntime/core/platform/windows/debug_alloc.cc]: @RyanUnderhill  FYI.

File: onnxruntime/core/session/ort_apis.h [hunk 6]
```diff
@@ -16,6 +16,8 @@ ORT_API(void, ReleaseTypeInfo, OrtTypeInfo*);
 ORT_API(void, ReleaseTensorTypeAndShapeInfo, OrtTensorTypeAndShapeInfo*);
 ORT_API(void, ReleaseSessionOptions, OrtSessionOptions*);
 ORT_API(void, ReleaseCustomOpDomain, OrtCustomOpDomain*);
+ORT_API(void, ReleaseMapTypeInfo, OrtMapTypeInfo*);
+ORT_API(void, ReleaseSequenceTypeInfo, OrtSequenceTypeInfo*);
```
Comment (snnn) [inline onnxruntime/core/session/ort_apis.h]: Please talk to @RyanUnderhill  offline. We can't allow such changes.

File: tools/ci_build/github/azure-pipelines/nuget/templates/cpu.yml [hunk 7]
```diff
@@ -9,7 +9,7 @@ jobs:
     AgentPool : 'Win-CPU-2019'
     ArtifactName: 'drop-nuget'
     JobName: 'Windows_CI_Dev'
-    BuildCommand:  '--build_dir $(Build.BinariesDirectory) --skip_submodule_sync --use_openmp --build_shared_lib --use_featurizers --enable_onnx_tests --use_telemetry --cmake_generator "Visual Studio 16 2019"'
+    BuildCommand:  '--build_dir $(Build.BinariesDirectory) --skip_submodule_sync --use_openmp --build_shared_lib --use_featurizers --enable_onnx_tests --use_telemetry --use_winml --cmake_generator "Visual Studio 16 2019"'
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/nuget/templates/cpu.yml]: I thought it will be a new pipeline and new package name for WinML?

File: tools/ci_build/github/azure-pipelines/win-gpu-ci-pipeline.yml [hunk 8]
```diff
@@ -42,7 +42,8 @@ jobs:
     displayName: 'Generate cmake config'
     inputs:
       scriptPath: '$(Build.SourcesDirectory)\tools\ci_build\build.py'
-      arguments: '--config $(BuildConfig) --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --update --cmake_generator "Visual Studio 16 2019" --msvc_toolset 14.16 --build_wheel --use_featurizers --use_dnnl --build_shared_lib --enable_onnx_tests --use_dml --use_cuda --cuda_version=10.0 --cuda_home="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.0" --cudnn_home="C:\local\cudnn-10.0-windows10-x64-v7.6.5.32\cuda" --cmake_extra_defines CMAKE_SYSTEM_VERSION=10.0.18362.0'
+      arguments: '--config $(BuildConfig) --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --update --cmake_generator "Visual Studio 16 2019" --msvc_toolset 14.16 --build_wheel --use_featurizers 
+      --use_dnnl --build_shared_lib --enable_onnx_tests --use_dml --use_winml --use_cuda --cuda_version=10.0 --cuda_home="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.0" --cudnn_home="C:\local\cudnn-10.0-windows10-x64-v7.6.5.32\cuda" --cmake_extra_defines CMAKE_SYSTEM_VERSION=10.0.18362.0'
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/win-gpu-ci-pipeline.yml]: Two lines?

File: winml/dll/module.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        failure.pszFile,
+        failure.pszFunction,
+        failure.uLineNumber);
+  }
+}
+
+extern "C" BOOL WINAPI DllMain(_In_ HINSTANCE hInstance, DWORD dwReason, _In_ void* lpvReserved) {
+  switch (dwReason) {
+    case DLL_PROCESS_ATTACH:
+      DisableThreadLibraryCalls(hInstance);
```
Comment (snnn) [inline winml/dll/module.cpp]: "DisableThreadLibraryCalls(hInstance)"
Is WinML magic statics free? If you have any magic static variables, this doesn't work.

**Comment by snnn** [comment]
Overall it LGTM.  There are some tiny things that must get fixed. 
You may exclude "mask_rcnn_keras" from the test cases to make your build pass. I'll take care it later.

File: BUILD.md [hunk 10]
```diff
@@ -84,6 +84,7 @@ For other system requirements and other dependencies, please see [this section](
 |**Build Shared Library**|--build_shared_lib||
 |**Build Python wheel**|--build_wheel||
 |**Build C# and C packages**|--build_csharp||
+|**Build WindowsML**|--use_winml<br>--use_dml<br>--build_shared_lib|WindowsML depends on DirectML and the OnnxRuntime shared library.|
```
Comment (pranavsharma) [inline BUILD.md]: Can we avoid users from typing --build_shared_lib and just infer it when --use_winml is specified? It's just one less thing to remember for users.

File: cmake/precompiled_header.cmake [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# Configures sources on a target to use a precompiled header. This function takes a target and
```
Comment (pranavsharma) [inline cmake/precompiled_header.cmake]: Please add license header and copyright notice.

File: cmake/wil.cmake [hunk 12]
```diff
@@ -0,0 +1,2 @@
+add_library(wil INTERFACE)
```
Comment (pranavsharma) [inline cmake/wil.cmake]: Please add license header and copyright notice.

File: cmake/winml_cppwinrt.cmake [hunk 13]
```diff
@@ -0,0 +1,220 @@
+# This script adds cppwinrt support for VS-generated projects.
```
Comment (pranavsharma) [inline cmake/winml_cppwinrt.cmake]: Please add license header and copyright notice.

File: cmake/winml_sdk_helpers.cmake [hunk 14]
```diff
@@ -0,0 +1,117 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (pranavsharma) [inline cmake/winml_sdk_helpers.cmake]: Please add license header and copyright notice.

File: docs/Privacy.md [hunk 15]
```diff
@@ -14,5 +14,5 @@ Currently telemetry is only implemented for Windows builds, but may be expanded
 #### Technical Details
 The Windows provider uses the [TraceLogging](https://docs.microsoft.com/en-us/windows/win32/tracelogging/trace-logging-about) API for its implementation.
 
-Windows ML and onnxruntime C APIs allow telemetry collection to be turned on/off (see API pages for details). It is turned off by default in all official builds.
+Windows ML and onnxruntime C APIs allow telemetry collection to be turned on/off (see API pages for details). In all Official builds, the Windows ML API enables telemetry collection.
```
Comment (pranavsharma) [inline docs/Privacy.md]: Please mention that it is turned OFF by default in the ORT APIs.

File: include/onnxruntime/core/common/exceptions.h [hunk 16]
```diff
@@ -32,6 +34,11 @@ class OnnxRuntimeException : public std::exception {
       : OnnxRuntimeException(location, nullptr, msg) {
   }
 
+  OnnxRuntimeException(const CodeLocation& location, const Status& status) noexcept
+      : OnnxRuntimeException(location, nullptr, status.ToString()) {
+    status_ = status;
```
Comment (pranavsharma) [inline include/onnxruntime/core/common/exceptions.h]: could be done in the initializer list?

File: onnxruntime/core/graph/schema_registry.cc [hunk 17]
```diff
@@ -194,6 +194,23 @@ DomainToVersionMap SchemaRegistryManager::GetLatestOpsetVersions(bool is_onnx_on
...
+    const std::string& domain,
+    const int op_set_version) {
+  // check the ONNX schema registry
+  auto& onnx_domain_version_map =
+      ONNX_NAMESPACE::OpSchemaRegistry::DomainToVersionRange::Instance().Map();
+
+  auto it = onnx_domain_version_map.find(domain);
+  if (it != onnx_domain_version_map.end() && op_set_version > it->second.second) {
+    // The domain is beyond what is registered.
+    return true;
```
Comment (pranavsharma) [inline onnxruntime/core/graph/schema_registry.cc]: ```return (it != onnx_domain_version_map.end() && op_set_version > it->second.second);``` should suffice, right?

File: onnxruntime/core/platform/telemetry.h [hunk 18]
```diff
@@ -10,6 +10,8 @@
 #include "core/common/status.h"
 #include "core/common/common.h"
 
+typedef struct _LUID LUID;
```
Comment (pranavsharma) [inline onnxruntime/core/platform/telemetry.h]: Can you use 'using' here?

File: onnxruntime/core/platform/windows/debug_alloc.cc (hunk 5 above)
Comment (pranavsharma) [reply onnxruntime/core/platform/windows/debug_alloc.cc]: What's the rationale for removing the message box? I found it pretty useful.

File: onnxruntime/core/session/inference_session.h [hunk 19]
```diff
@@ -512,6 +512,11 @@ class InferenceSession {
   const long long kDurationBetweenSending = 1000 * 1000 * 60 * 10;  // duration in (us).  send a report every 10 mins
   std::string event_name_;                                          // where the model is loaded from: ["model_loading_uri", "model_loading_proto", "model_loading_istream"]
 
+  TimePoint time_sent_last_evalutation_start_;
```
Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.h]: Would you mind grouping all the telemetry related members in a struct of it's own? It would make the class easier to read and maintain. This would be a good time to do it.

**Comment by pranavsharma** [comment]
I haven't checked all the files under the winml/ folder. Please make sure licence headers and copyright notices are added in all these new files.

File: winml/adapter/winml_adapter_c_api.h [hunk 20]
```diff
@@ -0,0 +1,469 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (pranavsharma) [inline winml/adapter/winml_adapter_c_api.h]: Can you document here or at an appropriate place that these APIs are an implementation detail to support the WinML APIs only and not intended for direct usage and that any direct usage of these APIs is not supported by the product? Just want to ensure any usage of these APIs don't turn into any support issues for us.

File: winml/dll/winml.rc [hunk 21]
```diff
@@ -0,0 +1,42 @@
+// This file REQUIRES the following external definitions:
```
Comment (pranavsharma) [inline winml/dll/winml.rc]: Are you going to do similar metadata changes for onnxruntime.dll as well or is that going to be a separate PR? cc Ori Levari

Comment (smk2007) [reply winml/dll/winml.rc]: Yeah, ill take them on today at some point and out a separate PR into master.

File: onnxruntime/core/platform/windows/debug_alloc.cc (hunk 5 above)
Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/debug_alloc.cc]: It links to desktop only libs. I removed all desktop only libs from the default link libraries and this was the only piece of code that required a desktop lib. This was an effort not to depend on desktop and to use onecoreuap libs.

We could conditionally link to desktop libs only on debug builds, but I thought it was easier to just remove this one message box.

Comment (snnn) [reply onnxruntime/core/platform/windows/debug_alloc.cc]: BTW, I'm trying to add a macro to control this memory leak checker:
https://github.com/microsoft/onnxruntime/commit/f85420fedbbb345e502e2b455129fa86cf1ed3f9#diff-83ba5c7244b867af5d86acd34e43f7ca

```c++
#ifdef ONNXRUNTIME_ENABLE_MEMLEAK_CHECK
```

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (smk2007) [reply cmake/CMakeLists.txt]: merge issue. will be resolved in next push

File: onnxruntime/core/graph/model.cc (hunk 3 above)
Comment (smk2007) [reply onnxruntime/core/graph/model.cc]: @fdwr Can you take a look at this feedback?

File: onnxruntime/core/graph/schema_registry.cc (hunk 4 above)
Comment (smk2007) [reply onnxruntime/core/graph/schema_registry.cc]: @fdwr Can you take a look at this feedback?

File: onnxruntime/core/platform/windows/debug_alloc.cc (hunk 5 above)
Comment (smk2007) [reply onnxruntime/core/platform/windows/debug_alloc.cc]: @tiagoshibata Can you please take a look at this feedback? It seems that the debugger present check may not be needed here?

File: tools/ci_build/github/azure-pipelines/nuget/templates/cpu.yml (hunk 7 above)
Comment (smk2007) [reply tools/ci_build/github/azure-pipelines/nuget/templates/cpu.yml]: AFAIK, winml will always be there in the windows packages created.

File: tools/ci_build/github/azure-pipelines/win-gpu-ci-pipeline.yml (hunk 8 above)
Comment (smk2007) [reply tools/ci_build/github/azure-pipelines/win-gpu-ci-pipeline.yml]: Back to 1 line.

File: cmake/precompiled_header.cmake (hunk 11 above)
Comment (smk2007) [reply cmake/precompiled_header.cmake]: Added

File: cmake/winml_cppwinrt.cmake (hunk 13 above)
Comment (smk2007) [reply cmake/winml_cppwinrt.cmake]: Added

File: include/onnxruntime/core/common/exceptions.h (hunk 16 above)
Comment (smk2007) [reply include/onnxruntime/core/common/exceptions.h]: Changes should not be included... this was reverted when the nano-com adapter was switched to C.

Removed.

File: onnxruntime/core/graph/schema_registry.cc (hunk 17 above)
Comment (smk2007) [reply onnxruntime/core/graph/schema_registry.cc]: @fdwr Can you take a look here too? Thanks!

File: onnxruntime/core/platform/telemetry.h (hunk 18 above)
Comment (smk2007) [reply onnxruntime/core/platform/telemetry.h]: done!

File: onnxruntime/core/session/inference_session.h (hunk 19 above)
Comment (smk2007) [reply onnxruntime/core/session/inference_session.h]: grouped into a struct

File: winml/adapter/winml_adapter_c_api.h (hunk 20 above)
Comment (smk2007) [reply winml/adapter/winml_adapter_c_api.h]: done

File: onnxruntime/core/platform/windows/debug_alloc.cc (hunk 5 above)
Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/debug_alloc.cc]: I didn't add the `IsDebuggerPresent`, it already existed. What was the rationale behind exiting early if a debugger is not present @RyanUnderhill ?

File: cmake/winml_sdk_helpers.cmake (hunk 14 above)
Comment (smk2007) [reply cmake/winml_sdk_helpers.cmake]: done

File: cmake/wil.cmake (hunk 12 above)
Comment (smk2007) [reply cmake/wil.cmake]: done

File: winml/dll/module.cpp (hunk 9 above)
Comment (smk2007) [reply winml/dll/module.cpp]: You are correct. We noticed and did a major statics cleanup a little while ago. there are still some exceptions, but we will resolve those statics soon.

File: onnxruntime/core/graph/schema_registry.cc (hunk 17 above)
Comment (fdwr) [reply onnxruntime/core/graph/schema_registry.cc]: @pranavsharma : Yep, equivalent and sufficient.

File: onnxruntime/core/graph/schema_registry.cc (hunk 4 above)
Comment (fdwr) [reply onnxruntime/core/graph/schema_registry.cc]: It really checks anything in this registry:

`OpSchemaRegistry::DomainToVersionRange&
OpSchemaRegistry::DomainToVersionRange::Instance() {
  static DomainToVersionRange domain_to_version_range;
  return domain_to_version_range;
};`

Which includes:
"com.microsoft.nchwc", (1, 1)
"", (1, 11)
"ai.onnx.ml", (1, 2)
"com.microsoft", (1, 1)
"com.microsoft.automl", (1, 1)
"com.microsoft.dml", (1, 1)

The comment is stale, because it really applies to any intrinsically known registered domain.

`// Either all ONNX domains were within range, or the domains were not ONNX.`

->

`// Either all domains were within range, or the domains were unknown.`

Unknown domains (e.g. "foo") will return false from this function (safer), which then checks `ONNX_NAMESPACE::OpSchemaRegistry::Schema` for a match.

File: onnxruntime/core/graph/model.cc (hunk 3 above)
Comment (fdwr) [reply onnxruntime/core/graph/model.cc]: Sounds fine to me. We can define a constant in `model.h` `kMaximumIrVersion` per my original commit, before the PR feedback to use `ONNX_NAMESPACE::Version::IR_VERSION` instead: https://github.com/microsoft/onnxruntime/pull/2413/commits/ae4dc20ec250a79c63339a1258a651f7b3e3abe1

Comment (snnn) [reply onnxruntime/core/graph/model.cc]: This isn't a blocking issue. We can have the discussion later(with Pranav).

File: onnxruntime/core/session/ort_apis.h (hunk 6 above)
Comment (smk2007) [reply onnxruntime/core/session/ort_apis.h]: yep yep!

File: cmake/CMakeLists.txt [hunk 22]
```diff
@@ -798,7 +798,7 @@ foreach(target_name onnxruntime_common onnxruntime_graph onnxruntime_framework o
 endforeach()
 
 foreach(provider_name ${ONNXRUNTIME_PROVIDER_NAMES})
-  if(NOT provider_name STREQUAL "cpu")
+  if(NOT provider_name STREQUAL "cpu" AND NOT provider_name STREQUAL "winml")
```
Comment (snnn) [inline cmake/CMakeLists.txt]: Which flag the winml doesn't support?

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml [hunk 23]
```diff
@@ -30,7 +30,7 @@ jobs:
     AgentPool : 'Win-GPU-2019'
     ArtifactName: drop-nuget-dml
     JobName: 'Windows_CI_GPU_DML_Dev'
-    BuildCommand: --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --use_featurizers --enable_onnx_tests --use_telemetry  --use_winml --use_dml --cmake_generator "Visual Studio 16 2019"
+    BuildCommand: --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --use_featurizers --enable_onnx_tests --use_telemetry --use_dml --cmake_generator "Visual Studio 16 2019"
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml]: use_winml?

File: cmake/CMakeLists.txt (hunk 22 above)
Comment (smk2007) [reply cmake/CMakeLists.txt]: onnxruntime_providers_winml doesnt have a target that gets built. Only headers.

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml (hunk 23 above)
Comment (smk2007) [reply tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml]: The lab has failing tests that need investigation so i have disabled gpu for now - once this bit change is in - i will be able to enable gpu in a more targetted fix.
